### PR TITLE
added checkpointing to support LLMs

### PR DESCRIPTION
**Pull request by hariharan-devarajan**
Changes to support Microsoft's Megatron Deepspeed.

- [x] Support for Checkpointing of model parameters, optimization states, and layer parameters.
- [x] Support for Reading Index dataset
- [x] Write configuration file megatron deepspeed.

### CODE REVIEW

**Comment by zhenghh04** [comment]
This PR addressed: https://github.com/argonne-lcf/dlio_benchmark/pull/88

**Comment by hariharan-devarajan** [comment]
@zhenghh04 I can confirm with the profiler that this change to checkpointing accurately represents the checkpointing in deepspeed. Additionally, the indexed_binary and mmap_indexed_binary are the two modes used in megatron deepspeed for data reading and the calls are accurate.

You can merge this if it looks good to u.

File: dlio_benchmark/framework/torch_framework.py [hunk 1]
```diff
@@ -97,19 +118,25 @@ def trace_object(self, string, step, r):
...
+            if self.optimization_state:
+                fname = os.path.join(self.checkpoint_folder, f"optimizer-{epoch}-{step_number}-{my_rank}.pt")
+                with open(fname, "wb") as f:
+                    torch.save(self.optimization_state, f)
+
+            if self.layer_state and self.args.num_layers > 0:
+                for layer in range(self.args.num_layers):
+                    fname = os.path.join(self.checkpoint_folder, f"layer-{layer}-{epoch}-{step_number}-{my_rank}.pt")
+                    with open(fname, "wb") as f:
+                        torch.save(self.layer_state, f)
```
Comment (zhenghh04) [inline dlio_benchmark/framework/torch_framework.py:139]: Make sure the conflict is solved.

**Review by zhenghh04 — CHANGES_REQUESTED** [review]

File: dlio_benchmark/configs/workload/dlrm.yaml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+reader:
+  data_loader: pytorch
+  batch_size: 2048
+  batch_size_eval: 16384
+  sample_shuffle: random
+
+train:
+  epochs: 1
+  computation_time: 0.064296
```
[the remark this answers is not in the document]
Comment (hariharan-devarajan) [reply dlio_benchmark/configs/workload/dlrm.yaml:28]: This is based on the configuration used in the PR #88

**Comment by hariharan-devarajan** [comment]
@zhenghh04 The original code uses TensorFlow and PyTorch APIs to save. This is needed as we are storing complex tensors. 

How would this work with S3? I think we need that fspec type interface for abstracting storage not manual abstraction. 

Thoughts?

File: dlio_benchmark/utils/config.py [hunk 3]
```diff
@@ -449,8 +452,16 @@ def LoadConfig(args, config):
...
+        if 'type' in config['checkpoint']:
+            args.checkpoint_type = CheckpointType(config['checkpoint']['type'])
         if 'model_size' in config['checkpoint']:
             args.model_size = config['checkpoint']['model_size']
+        if 'optimization_groups' in config['checkpoint']:
+            args.optimization_groups = config['checkpoint']['optimization_groups']
+        if 'num_layers' in config['checkpoint']:
+            args.num_layers = config['checkpoint']['num_layers']
+        if 'layer_parameters' in config['checkpoint']:
+            args.layer_parameters = config['checkpoint']['layer_parameters']
```
Comment (zhenghh04) [inline dlio_benchmark/utils/config.py:464]: Could you also add the corresponding documentation to doc/ for these input parameters?

**Review by zhenghh04 — CHANGES_REQUESTED** [review]
Thank you for all the implementation. This feature implemented here is very useful. Please address the issues raise up.

File: dlio_benchmark/common/enumerations.py [hunk 4]
```diff
@@ -17,6 +17,16 @@
 
 from enum import Enum
 
+class CheckpointType(Enum):
```
[the remark this answers is not in the document]
Comment (hariharan-devarajan) [reply dlio_benchmark/common/enumerations.py:20]: How about CheckpointIOType Just IOType might confuse with Reading.

File: dlio_benchmark/framework/tf_framework.py [hunk 5]
```diff
@@ -51,17 +53,41 @@ def __init__(self, profiling):
...
+        self.model_state = None
+        rank_to_checkpoint = self.args.my_rank
+        if self.args.checkpoint_type == CheckpointType.COLLECTIVE:
+            rank_to_checkpoint = 0
+        if rank_to_checkpoint == self.args.my_rank:
+            num_ranks = 1
+            if self.args.checkpoint_type == CheckpointType.COLLECTIVE:
+                num_ranks = self.args.comm_size
+            if self.args.model_size > 0:
+                self.model_state = {"a": self._get_tensor(self.args.model_size*num_ranks)}
```
[the remark this answers is not in the document]
Comment (hariharan-devarajan) [reply dlio_benchmark/framework/tf_framework.py:65]: model_size is size of model per GPU. 
We can define it as absolute model size of app in which case.
1. For Per GPU case we need to divide it.

Else if it is per GPU then we will have to multiply it for the Collective case.

File: dlio_benchmark/framework/tf_framework.py [hunk 6]
```diff
@@ -51,17 +53,41 @@ def __init__(self, profiling):
...
                 self.tensorboard = ProfilerFactory.get_profiler(Profiler.TENSORBOARD)
         self.reader_handler = None
+        self.model_state = None
+        rank_to_checkpoint = self.args.my_rank
+        if self.args.checkpoint_type == CheckpointType.COLLECTIVE:
+            rank_to_checkpoint = 0
+        if rank_to_checkpoint == self.args.my_rank:
+            num_ranks = 1
+            if self.args.checkpoint_type == CheckpointType.COLLECTIVE:
+                num_ranks = self.args.comm_size
```
[the remark this answers is not in the document]
Comment (hariharan-devarajan) [reply dlio_benchmark/framework/tf_framework.py:63]: Collective basically means in the context of checkpointing is that all data is collected by rank zero and written. I am open for a better word to describe it. Maybe Aggregated and Per-Process?

File: dlio_benchmark/data_generator/indexed_binary_generator.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        """
+        super().generate()
+        np.random.seed(10)
+        GB=1024**3
+        for i in dlp.iter(range(self.my_rank, int(self.total_files_to_generate), self.comm_size)):
+            dim1, dim2 = self.get_dimension()
+            sample_size = dim1 * dim2
+            total_size = sample_size * self.num_samples
+            write_size = total_size
+            MEMORY_SIZE = 2*GB
```
[the remark this answers is not in the document]
Comment (hariharan-devarajan) [reply dlio_benchmark/data_generator/indexed_binary_generator.py:58]: under dataset, I will add a configuration called generation_buffer_size. What do you think?

File: dlio_benchmark/configs/workload/dlrm.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (hariharan-devarajan) [reply dlio_benchmark/configs/workload/dlrm.yaml:28]: Agreed

File: dlio_benchmark/framework/torch_framework.py (hunk 1 above)
Comment (hariharan-devarajan) [reply dlio_benchmark/framework/torch_framework.py:140]: Fixed

File: dlio_benchmark/common/enumerations.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hariharan-devarajan) [reply dlio_benchmark/common/enumerations.py]: I named its CheckpointLocationType as RANK_ZERO or ALL_RANKS

File: dlio_benchmark/data_generator/indexed_binary_generator.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        return prefix_path + '.sz.idx'
+
+    @dlp.log
+    def generate(self):
+        """
+        Generator for creating data in NPZ format of 3d dataset.
+        """
+        super().generate()
+        np.random.seed(10)
+        GB=1024**3
```
[the remark this answers is not in the document]
Comment (hariharan-devarajan) [reply dlio_benchmark/data_generator/indexed_binary_generator.py]: FIXED

File: dlio_benchmark/framework/tf_framework.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (hariharan-devarajan) [reply dlio_benchmark/framework/tf_framework.py]: CALLED IT RANK_ZERO

File: dlio_benchmark/utils/config.py (hunk 3 above)
Comment (hariharan-devarajan) [reply dlio_benchmark/utils/config.py:465]: Done

File: dlio_benchmark/framework/tf_framework.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (hariharan-devarajan) [reply dlio_benchmark/framework/tf_framework.py:65]: Explained correctly in Doc

File: dlio_benchmark/data_generator/indexed_binary_generator.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (hariharan-devarajan) [reply dlio_benchmark/data_generator/indexed_binary_generator.py]: Done

**Review by zhenghh04 — APPROVED** [review]
This PR looks good now. 
But we need to validating DLRM and Magatron-Deepspeed config files. I'll create two issues to keep track of this.

**Pull request merged**